src/Provider/LinkedIn.php

```diff
@@ -173,7 +183,17 @@ protected function checkResponse(ResponseInterface $response, $data)
      */
     protected function createResourceOwner(array $response, AccessToken $token)
     {
-        return new LinkedInResourceOwner($response);
+        // If current accessToken is not authorized with r_emailaddress scope,
+        // getResourceOwnerEmail will throw LinkedInAccessDeniedException, it will be caught here,
+        // and then the email will be set to null
+        // When email is not available due to chosen scopes, other providers simply set it to null, let's do the same.
+        try {
```

Comment:
@OJezu - From a design perspective, I think this is a good approach. I am glad to see that you provided such a full solution even though I felt like you were apprehensive about doing that. This is great work!

The use case for which I don't see support is where the consumer _thinks_ their access token has the correct permissions and the resource owner email consistently comes up null because the provider swallowed the error that would've indicated that. The quality of the current solution that is helpful here is that the consuming application/developer can control when it tries to fetch the email and can catch related errors when they arise.

Do you think there is a justification for capturing that intent by overloading the `getResourceOwner` method? 

```php
    public function getResourceOwner(AccessToken $token, $getEmail = false)
    {
        $response = $this->fetchResourceOwnerDetails($token);

        return $this->createResourceOwner($response, $token, $getEmail);
    }

    protected function createResourceOwner(array $response, AccessToken $token, $getEmail = false)
    {
        $email = null;
        if ($getEmail) {
            $email = $this->getResourceOwnerEmail($token);
        }
        // etc
    }
```

I'm not advocating for that solution exactly, but I think what it offers the consuming app is the ability to programmatically control when that second call is made and can handle any exceptions that arise from it. Also, it is similar to your original solution design except the flag is not set on the provider; instead it is passed on the request itself.

I think it is fair to assume that many of the use cases for this library are the single-pass, access token was just freshly issued use case that you might find in SSO features. It is, however, possible that other consumers are using this library for other reasons where the access token have been stored previously and used when invoking this `getResourceOwner` method.

Reply:
The problem of users not getting some fields when invalid scopes are chosen by developer is same for all providers. E.g. if developer skips email scope for facebook adapter, they will get a null response from `getEmail` method, and no exception thrown IIRC. I don't see a solution for LinkedIn Provider that would be consistent with other Providers.

Adding an additional argument to `createResourceOwner` that has to be set to get an email address violates contract of the parent method - which is returning RO with all fields available in authorized scopes. That means, that generic code designed to handle `ProviderInterface` would not get email addresses from LinkedIn, unless it has a branch to specifically handle LinkedIn provider. That is why the first version had "getEmail" in Provider configuration - configuration is something that already differs between providers and had to be handled by existing code anyway.

I'm more concerned about developer being able to skip unnecessary query to LinkedIn, if they don't want email address, but that maybe such rare use case it can be ignored for now. I certainly don't need it, and unless a simple solution is found I'm not going to fret about it.

Reply:
> E.g. if developer skips email scope for facebook adapter, they will get a null response from getEmail method, and no exception thrown IIRC.

I feel like this is the only strong point you've made here. I think it's worth merging the change and getting it out there.

The rest of the points about "violates contract of the parent method", "generic code designed to handle ProviderInterface would not get email addresses from LinkedIn", and the like don't hold water with me. This package (and its many sibling providers) is built _upon_ the `oauth2-client` not built _for_ the `oauth2-client`. There are many cases where provider-specific implementations deviate or add to the stock functionality of the base provider. Expecting all implementations to maintain strict conformity to the base package is a fantasy - great if it were possible, but not a reality.